arangod/IResearch/IResearchLink.cpp

```diff
@@ -622,84 +633,160 @@ arangodb::Result IResearchLink::init(
+          TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
+          std::string("error finding view: '") + viewId + "' for link '" + std::to_string(_id) + "' : no such view"
+        );
+      }
+
+      auto* view = arangodb::LogicalView::cast<IResearchViewCoordinator>(
+        logicalView.get()
       );
+
+      if (!view) {
+        return arangodb::Result(
+          TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
+          std::string("error finding view: '") + viewId + "' for link '" + std::to_string(_id) + "'"
+        );
+      }
+
+      viewId = view->guid(); // ensue that this is a GUID (required by operator==(IResearchView))
+
+      if (!view->emplace(_collection.id(), _collection.name(), definition)) {
+        return arangodb::Result(
+          TRI_ERROR_INTERNAL,
+          std::string("failed to link with view '") + view->name() + "' while initializing link '" + std::to_string(_id) + "'"
+        );
+      }
     }
-  } else {
-    auto* view = LogicalView::cast<IResearchView>(logicalView.get());
+  } else if (arangodb::ServerState::instance()->isDBServer()) { // db-server link
+    auto* engine = arangodb::ClusterInfo::instance();
 
-    if (!view) {
+    if (!engine) {
       return arangodb::Result(
-        TRI_ERROR_ARANGO_DATA_SOURCE_NOT_FOUND,
-        std::string("error finding view: '") + viewId + "' for link '" + std::to_string(_id) + "'"
+        TRI_ERROR_INTERNAL,
+        std::string("failure to get storage engine while initializing arangosearch link '") + std::to_string(_id) + "'"
       );
     }
 
-    if (arangodb::ServerState::instance()->isDBServer()
-        && _collection.id() == _collection.planId()
-        && _collection.isAStub()) { // cluster cluster-wide link
-      auto shardIds = _collection.shardIds();
+    auto clusterWideLink = _collection.id() == _collection.planId() && _collection.isAStub(); // cluster cluster-wide link
 
-      // go through all shard IDs of the collection and try to link any links
-      // missing links will be populated when they are created in the per-shard collection
-      if (shardIds) {
-        for (auto& entry: *shardIds) {
-          auto collection = vocbase.lookupCollection(entry.first); // per-shard collections are always in 'vocbase'
+    if (!clusterWideLink) {
+      auto res = initDataStore(); // prepare data-store which can then update options via the IResearchView::link(...) call
 
-          if (!collection) {
-            continue; // missing collection should be created after Plan becomes Current
-          }
+      if (!res.ok()) {
+        return res;
+      }
+    }
+
+    auto logicalView = engine->getView(vocbase.name(), viewId); // valid to call ClusterInfo (initialized in ClusterFFeature::prepare()) even from Databasefeature::start()
+
+    // if there is no logicalView present yet then skip this step
+    if (logicalView) {
+      if (arangodb::iresearch::DATA_SOURCE_TYPE != logicalView->type()) {
+        unload(); // unlock the data store directory
```

Comment:
should `unload()` be called via a scopeGuard for exception safety ?

Reply:
propagating the exception up should be OK in this case